Item burning code rewrite, exploding magazines

Coolthulhu opened the pull request:
Closes #15239

A big rewrite of item burning code, which was an almost 200 line block of `if/else` cases.
Now material burning data is in jsons and items are responsible for burning themselves and returning how well did they burn.

This simplifies some mechanics and makes some materials easier to burn (for example, logs are no longer immune to small fires), but I don't think anyone cares that much about those formulas that were never all that accurate to begin with.

Ammo contained in a magazine can now explode, destroying the magazine in process.

Didn't test speed, but item burning code should now run faster due to removal of many `if( string == string )` comparisons in item materials and caching of ammo cookoff status in `item_factory` instead of determining it all at the time it burns.
Could be still optimized by precalculating per-item burn data instead of calculating it from 1-3 materials.

Fire code is still ugly, but the item part is now much cleaner.

codemime commented on src/item.cpp:
```diff
@@ -5142,13 +5191,72 @@ iteminfo::iteminfo(std::string Type, std::string Name, std::string Fmt,
+            if( content_destroyed ) {
+                content = contents.erase( content );
+            } else {
+                content++;
```
Could be shortened:
```c++
const auto new_end = std::remove_if( content.begin(), content.end(), [ &p, &drops ]( const item &it ) {
    return item.detonate( p, drops );
} );
if( new_end != content.end() ) {
    content.erase( new_end, content.end() );
    return true; // If any of the contents explodes, so does the container
}
```

mugling replied:
The above is more readable if it is correct?

codemime commented on src/item.cpp:
```diff
@@ -5142,13 +5191,72 @@ iteminfo::iteminfo(std::string Type, std::string Name, std::string Fmt,
-    g->explosion( p, type->explosion );
+    // Most containers do nothing to protect the contents from fire
+    if( type->magazine == nullptr || !type->magazine->protects_contents ) {
+        return any_of( contents.begin(), contents.end(), []( const item &it ) {
```
Missing namespace.

codemime commented on src/material.cpp:
```diff
@@ -84,6 +95,18 @@ void material_type::load_material( JsonObject &jsobj )
+            JsonObject brn = burn_data_array.next_object();
+            mat._burn_data[ intensity ] = load_mat_burn_data( brn );
+        } else {
+            // If not specified, supply default
```
Maybe use a previous data (if `intensiy` > 0).

Coolthulhu replied:
```
bool flammable = mat._fire_resist <= ( int )intensity;
```

covers more entries than copying or extrapolating would.

mugling commented on src/item_factory.cpp:
```diff
@@ -120,9 +120,32 @@ void Item_factory::finalize() {
+                    "TOXICGAS", "TEARGAS", "SMOKE", "SMOKE_BIG",
+                    "FRAG", "FLASHBANG"
+                }};
+                obj.ammo->special_cookoff = std::any_of( ammo_effects.begin(), ammo_effects.end(),
```
This is very readable syntax. Does it keep `astyle` happy?

Coolthulhu replied:
I don't recall it complaining. It might want to align the indents more agressively, possibly pushing `any_of` to next line.

mugling commented on src/itype.h:
```diff
@@ -444,6 +447,12 @@ struct islot_ammo : common_ranged_data {
      */
     int loudness = -1;
+
+    /** Should this ammo explode in fire? */
```
Needs comment explaining is cached via `item_factory`

mugling commented on src/fire.h:
```diff
@@ -0,0 +1,26 @@
+
+struct mat_burn_data {
+    // Made to protect other materials too
+    bool immune;
```
Defaults could go in the class rather than hidden in `item_factory`. The latter should probably use the `assign` syntax for consistency with the rest of that module.

Coolthulhu replied:
They're loaded in `material.cpp`, which doesn't use `assign`.

codemime commented on src/item.cpp:
```diff
@@ -2499,6 +2500,27 @@ int item::precise_unit_volume() const
+
 int item::volume( bool integral ) const
 {
     if( is_null() ) {
```
Since `item::base_volume()` duplicates a chunk of code from the `item::volume()` the latter should call the former.

mugling commented:
@Coolthulhu do you still want to work on this as the idea is good. I want to add some more materials so it would seem prudent to get this PR resolved first if so?

Coolthulhu commented:
I'll fix the issues in a moment. I'll push a hacky workaround for the segfault and then fix this one.

Coolthulhu commented:
I cleaned up partially:
* I couldn't find a good way to have `volume` use `base_volume` since they both want an early escape on nulls and corpses, so I just extracted the corpse volume switch to a new function
* Materials don't use assign, so I changed it to `read` and set up the defaults in `fire.h`
* Added a note about caching of cookoff `bool`s

mugling commented on src/material.cpp:
```diff
@@ -55,11 +55,11 @@ material_type::material_type()
-    bd.smoke = jsobj.get_int( "smoke", bd.smoke );
-    bd.burn = jsobj.get_int( "burn", bd.burn );
-    bd.chance_in_volume = jsobj.get_int( "chance", bd.chance_in_volume );
+    jsobj.read( "immune", bd.immune );
```
```
#include "generic_factory.h

...

assign( jsobj, "immune", bd.immune )
```

Coolthulhu replied:
Doesn't it also require some extra setting up, like creating a factory for the type?

mugling replied:
Nope. It's a straight wrapper around `read` but with some added magic. For example it can parse arrays of `string_id<foo>`. The plan being to keep `json.h` lightweight but add higher-order functions separately.

mugling commented:
Acetylene tanks don't explode
Ammo explodes too much (a STANAG magazine explodes multiple times each with significant damage)

mugling commented on src/item.cpp:
```diff
@@ -2486,24 +2487,42 @@ int item::precise_unit_volume() const
 }
 
-int item::volume( bool integral ) const
+int corpse_volume( m_size corpse_size )
```
could be static function

mugling commented on src/item.cpp:
```diff
@@ -5138,13 +5184,72 @@ iteminfo::iteminfo(std::string Type, std::string Name, std::string Fmt,
 
-    g->explosion( p, type->explosion );
+    // Most containers do nothing to protect the contents from fire
+    if( type->magazine == nullptr || !type->magazine->protects_contents ) {
```
`is_magazine()`

Coolthulhu commented:
Acetylene tanks exploding would require giving acetylene combat stats and removing hydrocarbon material or special casing. "COOKOFF" effect can't do anything without combat stats.
Should it get combat stats?

mugling commented on src/itype.h:
```diff
@@ -420,7 +420,10 @@ struct islot_magazine {
+    itype_id linkage = "NULL";
+
+    /** If false, ammo will cook off if this mag is affected by fire */
+    bool protects_contents = false;
```
Better named `insulated`. Eventually this should be a scalar as opposed to boolean value specifying the degree of protection.

Coolthulhu replied:
Insulated sounds a bit too electric-related

mugling replied:
`thermal_insulation`

Coolthulhu replied:
Could be

Though currently it protects from all kinds of damage that goes through mags and into contents (only fire == all kinds).
If there are more kinds later on, it would work better as a map of `type->protection` values.

mugling commented on src/itype.h:
```diff
@@ -446,6 +449,20 @@ struct islot_ammo : common_ranged_data {
+    /**
+     * Should this ammo explode in fire?
+     * This value is cached by item_factory based on ammo_effects and item material.
+     * It is not read from the json directly.
```
Markup for doygen:
`@warning It is not read from the json directly`

mugling commented on src/item.cpp:
```diff
@@ -2486,24 +2487,42 @@ int item::precise_unit_volume() const
+    return 0;
+}
+
+int item::base_volume() const
```
If performance is important could be an inline function in `item.h` as
`return is_corpse() ? corpse_volume( corpse->size ) : type->volume`

mugling replied:
Alternatively if the fire code is the only consumer just move this logic within the fire code and add suitable documentation

mugling commented:
> Acetylene tanks exploding would require giving acetylene combat stats and removing hydrocarbon material or special casing. "COOKOFF" effect can't do anything without combat stats.
> Should it get combat stats?

No, that would be obtuse. Could the burn data not be extended with an additional field for materials that are extrinsically explosive?

Coolthulhu commented:
> Could the burn data not be extended with an additional field for materials that are extrinsically explosive?

I just remembered all objects have explosion data and explode in fire flag, will try that way.

mugling commented on src/item_factory.cpp:
```diff
@@ -129,9 +129,32 @@ void Item_factory::finalize() {
+            if( std::find( mats.begin(), mats.end(), material_id( "hydrocarbons" ) ) == mats.end() &&
+                std::find( mats.begin(), mats.end(), material_id( "oil" ) ) == mats.end() ) {
+                const auto &ammo_effects = obj.ammo->ammo_effects;
+                obj.ammo->cookoff = ammo_effects.count( "INCENDIARY" ) > 0 ||
```
`ammo_effects` is a `std::set` and can only ever contain zero or one of an element which this code does not imply. STL is deficient and should really provide an `exists` or `has` method but as it doesn't consider:

`ammo_effects.count( "INCENDIARY" )`

Coolthulhu replied:
It's cleaner with an explicit comparison rather than an implicit cast to bool.

mugling replied:
It implies that `ammo_effects` could contain `INCENDIARY` multiple times

Coolthulhu replied:
That's implied by `std::set` return value, not the code that uses it.

mugling commented:
> I just remembered all objects have explosion data and explode in fire flag, will try that way.
> 
> EDIT: It would make it not scale with charges, though

Better than both not exploding at all or gaining unused stats

mugling commented on data/json/items/ammo/weldgas.json:
```diff
@@ -15,6 +15,11 @@
-    "flags": [ "NO_DROP", "IRREMOVABLE" ]
+    "flags": [ "NO_DROP", "IRREMOVABLE" ],
+    "explode_in_fire" : true,
+    "explosion" : {
```
Having an `explosion` member should automatically imply `explodes_in_fire` which could be achieved by relatively simple logic in `item_factory.cpp`

Coolthulhu replied:
Not necessarily
At the moment it is true, but that's only because things like chain explosions aren't in and because all `explosion_iuse` tools specify their own explosions inside the actor. There are objects that should only explode under specific conditions, such as C4 and ANFO. They could explode in raging fire, but not in density 1 one.

mugling commented on data/json/items/ammo/weldgas.json:
```diff
@@ -15,6 +15,11 @@
+    "explosion" : {
+        "power": 20,
+        "shrapnel": 0
+    }
```
Style for the `ammo` definitions always includes `flags` as the last element

mugling commented on src/item.cpp:
```diff
@@ -4724,15 +4743,50 @@ bool item::reload( player &u, item_location loc, long qty )
+    int time_added = 0;
+    int burn_added = 0;
+    const int vol = base_volume();
+    for( const auto &mat : mats ) {
```
`mat` vs` mats` is a slightly evil naming scheme.

```
for( const auto &e : mats ) {
}
```

mugling commented:
Acetylene now works including when a tank is contained by a tool but is is too weak for a full tank. Is there any way it could be scaled with charges?

Coolthulhu commented:
> Is there any way it could be scaled with charges?

Would require a whole new way to set it up, not just a trick.
